File: backend/src/Domino.Backend/Utilities/ServiceResult.cs
namespace Domino.Backend.Utilities;

public enum ServiceErrorKind
{
    NotFound,
    Validation
}

public record ServiceError(ServiceErrorKind Kind, string Message);

public static class ServiceResult
{
    public static ServiceResult<T> Success<T>(T value) => new(value);
    public static ServiceResult<T> NotFound<T>(string message) => new(new ServiceError(ServiceErrorKind.NotFound, message));
    public static ServiceResult<T> Invalid<T>(string message) => new(new ServiceError(ServiceErrorKind.Validation, message));
}

public sealed class ServiceResult<T>
{
    public T? Value { get; }
    public ServiceError? Error { get; }
    public bool IsSuccess => Error is null;

    internal ServiceResult(T value) => Value = value;
    internal ServiceResult(ServiceError error) => Error = error;
}

File: backend-ts/src/common/service-result.ts
export type ServiceResult<T> =
  | { kind: 'success'; value: T }
  | { kind: 'not_found'; message: string }
  | { kind: 'invalid'; message: string };

export function success<T>(value: T): ServiceResult<T> {
  return { kind: 'success', value };
}

export function notFound<T>(message: string): ServiceResult<T> {
  return { kind: 'not_found', message };
}

export function invalid<T>(message: string): ServiceResult<T> {
  return { kind: 'invalid', message };
}
